### PR TITLE
Fix RoPE position inconsistencies across chunked samples

### PR DESCRIPTION
**Pull request by francesco-bertolotti**
### Summary

This PR removes the modulo-based logic previously used to wrap RoPE positions:

```python
self._positions_buffer.extend(
    i % self.seq_len for i in range(len(sample_tokens))
)
```

Although this prevented RoPE indices from exceeding the cache size, it introduced inconsistencies in positional encoding across split samples.

---

### Problem

Using modulo to reset positions can break positional continuity when a single logical sequence is split across multiple chunks.

For example:

```text
tokens:    [tx ,ty ,tz ,t0 ,t1 ,t2 ,t3 ,t4 ,t5 ,t6 ,t7 ]
positions: [x  ,y  ,z  ,0  ,1  ,2  ,3  ,4  ,5  ,6  ,7  ]

tokens:    [t8 ,t9 ,t10,t11,t12,t13,t14,t0 ,t1 ,t2 ,t3 ]
positions: [8  ,9  ,10 ,0  ,1  ,2  ,3  ,0  ,1  ,2  ,3  ]
```

Here, the sequence `t0–t14` is split across two chunks. In the second chunk, positions are reset mid-sequence due to the modulo operation,even though the tokens belong to the same original sequence. This results in inconsistent positional encodings.

Since different samples in a batch cannot attend to each other, positional resets should align strictly with sample boundaries, not occur within a continued sequence.

---

#### 1. Removed modulo-based position wrapping

Eliminated `% seq_len` to preserve positional continuity within sequences.

#### 2. Introduced `_normalize_positions`

Adds a normalization step that ensures each sequence starts at position 0. This is applied in the dataloader.

#### 3. Added position validation tests

New test verify that:

* Positions are monotonically increasing within a sequence
* Positions reset to 0 after an EOS token
* Each sequence starts at position 0
* BOS tokens are always assigned position 0
* Positions remain within valid bounds (≤ sequence length)

### CODE REVIEW

File: torchtitan/hf_datasets/text_datasets.py [hunk 1]
```diff
@@ -122,28 +150,26 @@ def __iter__(self):
...
-                # positions at 0 per request.  Positions wrap at seq_len
-                # to stay within the RoPE cache, effectively chunking
-                # long documents into seq_len-sized segments.
-                # TODO: make overflow policy configurable (chunk / truncate / drop).
-                self._positions_buffer.extend(
-                    i % self.seq_len for i in range(len(sample_tokens))
-                )
                 self._sample_idx += 1
 
+                sample_tokens = self._apply_overflow_policy(sample_tokens, max_buffer_token_len)
```
Comment (tianyu-l) [inline torchtitan/hf_datasets/text_datasets.py]: This doesn't sound right in terms of how the `overflow_policy` works with packing. Here's an example:
- `sample_tokens` is of length 90.
- `max_buffer_token_len` is of length 100.
- `self._inputs_buffer`, before adding this particular `sample_tokens`, already has length 99.

According to the policy, we would still get full `sample_tokens` because 90 < 100. However, it will be truncated below, regardless of the policy.

File: torchtitan/hf_datasets/text_datasets.py [hunk 2]
```diff
@@ -211,6 +237,18 @@ class Config(ParallelAwareDataloader.Config):
         infinite: bool = True
         """Whether to loop the dataset infinitely"""
 
+        overflow_policy: Literal["chunk", "truncate", "drop"] = "chunk"
+        """
+        Policy for handling samples that exceed seq_len: 
+        'chunk' (default) splits into multiple samples, 'truncate' truncates to fit, and 'drop' skips the sample.
+        """
+
+        reset_positions_on_chunk: bool = True
```
Comment (tianyu-l) [inline torchtitan/hf_datasets/text_datasets.py]: I'm worried about out-of-index error if we set it to False, and we get an ultra-long sequence whose length is more than `training.seq_len`.
https://github.com/pytorch/torchtitan/blob/main/torchtitan/models/common/rope.py#L232

Maybe we can use `torch._assert_async` to do assertion there.

**Comment by joecummings** [comment]
Thanks for being proactive about this one - we discussed it briefly in #2769 / #2757. 

I'd like to get this landed but simplified.
1. Can we move the drop/truncate/chunk to a new PR? That's not technically part of this fix. Leave the TODO in there
2. I think we can let the positions buffer extend monotonically with `self._positions_buffer.extend(range(len(sample_tokens)))` and then normalize the leading document fragement if the first position id AFTER slicing is > 0.

File: torchtitan/hf_datasets/text_datasets.py (hunk 2 above)
Comment (francesco-bertolotti) [reply torchtitan/hf_datasets/text_datasets.py]: I kept this optional because some users may want direct access to the true token positions within the model and handle positional embeddings using a custom implementation. That said, keeping customization simple might be preferable, especially since it’s straightforward to extend `HuggingFaceTextDataset` and override `_normalize_positions` if needed.

**Comment by francesco-bertolotti** [comment]
Happy to help @joecummings.
> Can we move the drop/truncate/chunk to a new PR? That's not technically part of this fix. Leave the TODO in there.

No problem, as @tianyu-l pointed out, the current implementation was a bit lacking anyway.

> I think we can let the positions buffer extend monotonically with `self._positions_buffer.extend(range(len(sample_tokens)))` and then normalize the leading document fragment if the first position id *after slicing* is > 0.

In the latest commit, `_normalize_positions` implements exactly this behavior.

**Comment by francesco-bertolotti** [comment]
@joecummings, thanks for the feedback, I have updated the PR description and added a test regarding positions.

**Comment by francesco-bertolotti** [comment]
I don’t think the linting error is related to this PR, as it appears to be tied to the training and validation loss stack. I’ll take a closer look at it tomorrow.

**Comment by tianyu-l** [comment]
oh might be fixed in https://github.com/pytorch/torchtitan/commit/3057525dc6a88e3c8fe5d961530989fea39b9183
